packages/devtools_app/lib/src/charts/chart_trace.dart

```diff
@@ -0,0 +1,213 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math';
+
+import 'package:flutter/material.dart';
+
+import 'chart_controller.dart';
+
+class Data {
+  Data(this.timestamp, this.y);
+
+  final int timestamp;
+  final double y;
+}
+
+class PaintCharacteristics {
+  PaintCharacteristics({
+    this.color = Colors.black,
+    this.symbol = ChartSymbol.ring,
+    this.strokeWidth = 1,
+    this.diameter = 1,
+    this.fixedMinY,
+    this.fixedMaxY,
+  });
+
+  /// If specified Y scale is computed and min value is fixed.
+  /// Will assert if new data point is less than min specified.
+  double fixedMinY;
+
+  /// If specified Y scale is computed and max value is fixed.
+  /// Will assert if new data point is more than max specified.
+  double fixedMaxY;
+
+  Color color;
+
+  ChartSymbol symbol;
+
+  double strokeWidth;
+
+  double diameter;
+}
+
+class Trace {
+  Trace(this.controller, this._chartType, this.characteristics) {
+    double minY = characteristics.fixedMinY;
+    minY ??= 0.0;
+
+    double maxY = characteristics.fixedMaxY;
+    maxY ??= 0.0;
+
+    yAxis = AxisScale(minY, maxY, 30);
+  }
+
+  final ChartController controller;
+
+  final ChartType _chartType;
+
+  final PaintCharacteristics characteristics;
+
+  String name;
+
+  double dataYMax = 0;
+
+  final data = <Data>[];
+
+  ChartType get chartType => _chartType;
+
+  AxisScale yAxis;
+
+  void addDatum(Data datum) {
+    data.add(datum);
+
+    if (characteristics.fixedMaxY != null) {
+      assert(
+        datum.y <= characteristics.fixedMaxY,
+        'y=${datum.y} fixedMaxY=${characteristics.fixedMaxY}',
+      );
+    } else if (datum.y > dataYMax) {
+      dataYMax = datum.y.toDouble();
+      yAxis = AxisScale(0, dataYMax, 30);
+    }
+
+    if (datum.y > controller?.yMaxValue) controller?.yMaxValue = datum.y;
+
+    final traceIndex = controller.traceIndex(this);
+
+    // New data has arrived notify listeners this data needs to be plotted.
+    controller.traceChanged.value = TraceNotifier(traceIndex, data.length - 1);
+  }
+}
+
+class TraceNotifier {
```

Comment:
document what traceIndex and dataIndex are.

Reply:
Done.